adds support for configuring commands and channels using YAML

elmjag opened the pull request:
Adds support for configuring commands and channels using YAML. Currently it's possible to configure `Tango`, `exporter` and `EPICS` commands and channels.

This implements roughly the format discussed in: https://github.com/mxcube/mxcubecore/issues/1023
The actual format implement is documented here: https://mxcubecore--1034.org.readthedocs.build/en/1034/dev/commands_channels.html

This is still work in progress. Still on the todo list:

* write some tests
* do a bit more "manual" testing

I would like to get some feedback, before spending more time on this.

Commit:
add docs on commands and channels YAML configuration format

Adds a section the describes how to configure hardware object's
commands and channels using YAML configuration files.

Documents the format for Tango, exporter and EPICS protocols.

Commit:
adds support for configuring commands and channels using YAML

Adds support for 'tango', 'exporter' and 'epics' sections to YAML
configuration files.

These section are used to configure Command and Channel objects for
hardware objects.

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
```diff
@@ -0,0 +1,288 @@
+| property  | purpose                           | default             |
+|-----------|-----------------------------------|---------------------|
+| attribute | tango attribute name              | MXCuBE channel name |
+| poll      | polling periodicity, milliseconds | polling is disabled |
```
```suggestion
| polling_period      | polling periodicity, milliseconds | polling is disabled |
```

elmjag replied:
changed to `polling_period`

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
```diff
@@ -0,0 +1,288 @@
+
+The mxcubecore provides a hardware object-level abstraction
+for communicating using various flavors of control system protocols.
+A hardware object can utilize instances of `Command` and `Channel` objects.
```
Link to the documentation for the actual classes, please.
(I always forget how to do it, look up the Myst doc: <https://myst-parser.readthedocs.io/en/latest/syntax/cross-referencing.html>).

Also isn't it `mxcubecore.CommandContainer.ChannelObject` and `mxcubecore.CommandContainer.CommandObject`?

elmjag replied:
Yes, it's `mxcubecore.CommandContainer.ChannelObject` and `mxcubecore.CommandContainer.ChannelObject`. Added links to API docs for these objects.

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
```diff
@@ -0,0 +1,288 @@
+for communicating using various flavors of control system protocols.
+A hardware object can utilize instances of `Command` and `Channel` objects.
+These objects provide a uniform API for accessing a specific control system.
+Mxcubecore provides support for using protocols such as _tango_, _EPICS_, _exporter_ and more.
```
Links, links, links! Documentation should have lots of links. Typically the first occurrence of every term should have a link to its definition.

```suggestion
Mxcubecore provides support for using protocols such as [_tango_](https://pytango.readthedocs.io/), _EPICS_, _exporter_ and more.
```

henrique-silva replied:
https://docs.epics-controls.org/en/latest/ for the official EPICS docs (that is being reworked now as far as I know, but at least it's not the APS website from the 80's anymore 😄)

elmjag replied:
Added links to Tango and EPICS web sites.

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
````diff
@@ -0,0 +1,288 @@
+The general format for specifying `Command` and `Channel` objects is as follows:
+
+```yaml
+class: MegaCommunicator
````
To keep consistent with the rest of the example:

```suggestion
class: <hardware-object-class>
```

elmjag replied:
changed

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
````diff
@@ -0,0 +1,288 @@
+The format for specifying _tango_ `Command` and `Channel` objects is as follows:
+
+```yaml
+class: TangoCommunicator
````
```suggestion
class: <hardware-object-class>
```

elmjag replied:
changed

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
```diff
@@ -0,0 +1,288 @@
+| property  | purpose                           | default             |
+|-----------|-----------------------------------|---------------------|
+| attribute | tango attribute name              | MXCuBE channel name |
+| poll      | polling periodicity, milliseconds | polling is disabled |
```
I guess there should be `timeout` as well, no? And `username`? Well, but only if anyone actually uses those, otherwise better leave them out indeed.

elmjag replied:
For `username`, I think it's should be deprecated. Looking that the code, I don't see any practical uses of the `username` anywhere. Feels like it's a left over from some older feature, which is not used anymore.

For `timeout`, I'll try to add it. That seems like a useful feature, as sometimes you need to tweak tango timeout.

elmjag replied:
I have added `timeout` config.

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
````diff
@@ -0,0 +1,288 @@
+Below is an example of a hardware object that specifies Tango commands and channels.
+
+```yaml
+class: MyTango
````
```suggestion
class: my_package.my_sub_package.MyTangoClass
```

Otherwise that is a good example: complex and complete. That is good.

elmjag replied:
But this is not an example of how to write a hardware configuration file. This is an example how to configure tango commands and channels. This long class name will distract the reader from the main point of the example.

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
````diff
@@ -0,0 +1,288 @@
+          poll: 1024
+```
+
+In the above example, commands `Open`, `Close` and `Reset` as well as `State` and `Volume` channels are configured.
````
```suggestion
In the above example, commands `Open`, `Close` and `Reset` as well as channels `State` and `Volume` are configured.
```

elmjag replied:
fixed

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
````diff
@@ -0,0 +1,288 @@
+```
+
+In the above example, commands `Open`, `Close` and `Reset` as well as `State` and `Volume` channels are configured.
+All command and channel objects are bound to the commands and attributes of the _some/tango/device_ tango device.
````
```suggestion
All command and channel objects are bound to the commands and attributes of the `some/tango/device` tango device.
```

elmjag replied:
This is to make it consistent with the style below.

https://github.com/mxcube/mxcubecore/pull/1034#discussion_r1798848269

fabcor-maxiv commented on docs/source/dev/commands_channels.md:
```diff
@@ -0,0 +1,288 @@
+In the above example, commands `Open`, `Close` and `Reset` as well as `State` and `Volume` channels are configured.
+All command and channel objects are bound to the commands and attributes of the _some/tango/device_ tango device.
+
+`Open` and `Close` commands are bound to _Open_ and _Close_ Tango commands.
```
```suggestion
`Open` and `Close` hardware object commands are bound to `Open` and `Close` Tango commands.
```

elmjag replied:
Here the italics mark-up is used to distinguish between Open in the context of the MXCuBE configuration and Open in the context of the tango device.

fabcor-maxiv commented on docs/source/dev/configuration_files.md:
```diff
@@ -59,18 +59,27 @@ and an `init()` function that is executed after configured parameters and contai
+  darkhan: darkhan.yaml
   session: session.yaml
+tango:
+  "some/tango/device":
```
```suggestion
  some/tango/device:
```

elmjag replied:
changed

fabcor-maxiv commented on mxcubecore/model/protocols/epics.py:
We should try to follow the Google style for doctrings:

* <https://google.github.io/styleguide/pyguide.html#38-comments-and-docstrings>
* <https://www.sphinx-doc.org/en/master/usage/extensions/example_google.html>

elmjag replied:
but it's so ugly :(

fabcor-maxiv commented on mxcubecore/model/protocols/epics.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class Channel(BaseModel):
+    """
+    an EPICS channel configuration
+    """
```
```suggestion
    """EPICS channel configuration."""
```

elmjag replied:
reluctantly changed!

fabcor-maxiv commented on mxcubecore/model/protocols/epics.py:
```diff
@@ -0,0 +1,57 @@
+"""
+Models the `epics` section of YAML hardware configuration file.
+
+Provides an API to read configured EPICS channels.
+"""
```
```suggestion
"""Models for the `epics` section of YAML hardware configuration file.

Provides an API to read configured EPICS channels.
"""
```

elmjag replied:
stupid google!

fabcor-maxiv commented on mxcubecore/model/protocols/tango.py:
```diff
@@ -0,0 +1,84 @@
+    __root__: Dict[str, Device]
+
+    def get_tango_devices(self) -> Iterable[Tuple[str, Device]]:
+        """
+        get all tango devices specified in this section
+        """
```
```suggestion
        """Get all Tango devices specified in this section."""
```

elmjag replied:
![image](https://github.com/user-attachments/assets/dc4b23b6-1dda-432e-b67d-b10430519071)

fabcor-maxiv commented on mxcubecore/protocols_config.py:
```diff
@@ -0,0 +1,135 @@
+    """
+    add the Command and Channel objects configured in the specified protocol section
+
+    parameters:
```
I think the sections are case-sensitive.

```suggestion
    Parameters:
```

elmjag replied:
works fine with lower case `p`:

![image](https://github.com/user-attachments/assets/04a6d6c7-5cc5-4c64-8091-67c6c7f808c1)

fabcor-maxiv commented on mxcubecore/protocols_config.py:
```diff
@@ -0,0 +1,135 @@
+    protocol_config = config.get(protocol)
+    if protocol_config is None:
+        # no configuration for this protocol
+        return
```
Why not raise an exception instead of failing silently?

Or am I misunderstanding something?

elmjag replied:
This is not an error.

This is the case when the config does not have a section for that specific protocol. I.e. there are no commands/channels configured using that protocol. This is normal situation.

The algorithm for handling protocol sections is as follows:

```
for protocol in <supported-protocols>:
  if config have protocol section then
    handle that protocol section
  else
   do nothing
```

fabcor-maxiv replied:
Ah I see : )

~I guess I would expect the "for" and "if" in one function, and the "handle" in the other function. But maybe that is not possible.~

fabcor-maxiv reviewed (commented):
Good stuff.

A first review pass, a first batch of comments and improvement suggestions.

henrique-silva commented on docs/source/dev/commands_channels.md:
````diff
@@ -0,0 +1,288 @@
+If no suffix is specified, the channel's MXCuBE name is used in place of the `suffix`.
+Observe an example configuration below:
+
+```yaml
+class: EpicsCommunicator
+epics:
+  "FOO:B:x1.":
+    channels:
+      State:
+        suffix: stat
+      Vol:
+```
+
+We have an `FOO:B:x1.` prefix specified, with two `Channel` objects `State` and `Vol`.
+`State` will use `FOO:B:x1.stat` PV name, specified by section's prefix and the `suffix` configuration property.
+`Vol` will use `FOO:B:x1.Vol` PV name, specified by section's prefix and channels MXCuBE name.
````
So, I think this might be a bit problematic, because the modeling of PV is usually `$(PREFIX):$(DEVICE).$(PROPERTY)`, so one record - or PV - ( `$(PREFIX):$(DEVICE)`) will have multiple properties, for example an analog out has, .DESC (description), .VAL (value), .ASLO (slope).
But in this case that you're modelling it is way more likely that this will be separated into multiple PVs, so you'd likely have:

```yaml
class: EpicsCommunicator
epics:
  "FOO:B:":
    channels:
      State:
        suffix: pv_1.STAT
      Vol:
        suffix: volume.VAL
```

In the end I think the code will handle just fine if you're just appending, but the prefix is only the part before teh last colon (in most naming conventions of EPICS that I've seen)

fabcor-maxiv replied:
I am not sure I am understanding you correctly, but would <https://github.com/mxcube/mxcubecore/issues/1023#issuecomment-2367904138> answer your question?

henrique-silva replied:
Yes, that's pretty much it.

It's just the non-obvious separation between PV/record and properties in EPICS that is a bit weird for modelling, but the simple concatenation should handle all cases

fabcor-maxiv replied:
@elmjag I think it is worth modifying the epics example to show a more complex and complete example, that shows how to combine all those things.

elmjag replied:
Will this example make things more clear?

```
class: EpicsCommunicator
epics:
  "FOO:B:":
    channels:
      State:
        suffix: pv_1.STAT
      Vol:
        suffix: volume.VAL
      Freq:
```

Regarding `Freq` channel, I want to include an example channel, without the optional config property `suffix`.

henrique-silva replied:
Yeah, I think it's clearer, but take my opinion with a grain of salt since I don't really know much about MxCube 😅 

But yes, it does show that you can map PVs and fields to the mxcube channels

elmjag replied:
updated example

Commit:
restructure slightly the docs strings in CommandContainer module

This way, it's possible to link to generated API documentation for
CommandContainer and ChannelObject classes.

Without this change, sphix refuses to create a link for
CommandContainer and ChannelObject classes, from other section of
the documentation.

fabcor-maxiv commented on mxcubecore/model/protocols/epics.py:
```diff
@@ -0,0 +1,47 @@
+
+    channels: Optional[Dict[str, Optional[Channel]]]
+
+    def get_channels(self) -> Iterable[Tuple[str, Channel]]:
```
I think it should be `Iterator` and not `Iterable`.

```suggestion
    def get_channels(self) -> Iterator[Tuple[str, Channel]]:
```

elmjag replied:
It is valid, generators _are_ Iterables:

```
>>> from typing import Iterable
>>> def foo():
...   yield 1
...   yield 3
>>> isinstance(foo(), Iterable)
True
```

Here I want to specify a more generic return type. It is _some_ kind of iterable. Today it's generator, but in the future perhaps it is a list.

fabcor-maxiv replied:
I see, I guess it is fine... I do not know for sure.

---

The rules I had read some time ago were like this:

* When you produce you should use as precise a type you can.
* When you consume you should use as generic a type as you can that still fulfils your usage.

so that is what I always follow and I never had any issue.

So for example, I would do something like this:

```python
def producer() -> list[int]:
    return [1, 2, 3]

def consumer(things: typing.Iterable[int]) -> None
    for thing in things:
        print(thing)

some_things = producer()
consumer(some_things)
```

If later the `producer` function is refactored into a generator (or iterator or any other iterable), I still do not need to change anything in `consumer`.

rhfogh replied:
A question purely for my education (you guys have studied more than I have). Reading you example one might be tempted to think that producer promises to return a list. Would it be OK to write code like 

list2 = list1 + producer() 

Or should one  instead write

list2 = list1 + list(producer())

in case producer was later refactored to return something else? Or is this principle only valid for certain contexts?

(I know you could write something like 

list2 = []; list2.extend(list1); list2.extend(producer()) 

but the question is whether you need to assume that the return type of producer() might be(come) any iterable at some later date.

elmjag replied:
If you have

```
def producer() -> list[int]:
    return [1, 2, 3]
```

Then, it explicitly promises to return a list. Thus you can use return value as a list.

However, if you have:

```
def producer() -> Iterable[int]:
    return [1, 2, 3]
```

Then it does not promise a list. This can any type of iterable value. Then you should explicitly convert this to a list, if you a doing list operations on it.

Here is documentation on Iterable type annotation: https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable


However, bear in mind that none of this is enforced during runtime. Nothing in python stops you from doing this:

```
def producer() -> Iterable[int]:
    return False
```

Python will happily execute this and return a boolean to you.

The only way to find you that your code is not following the type annotations is to run a static checker tool, such as mypy.

rhfogh replied:
@elmjag Thanks, that makes sense. I know that Python does not respect type annotations, but I want to be a good boy and follow our rules, so I was trying to make sure I understood what our rules were.

marcus-oscarsson replied:
I suggest that we try out pyright: https://github.com/microsoft/pyright

elmjag replied:
> I suggest that we try out pyright: https://github.com/microsoft/pyright

Yes, I agree. I wrote a new issue: https://github.com/mxcube/mxcubecore/issues/1053
